### ISSUE TITLE
Curious about the training set

### ISSUE THREAD

**Issue by PeterPanUnderhill** [bfelbo/DeepMoji#22]
I came across this question when reading the paper, and I cannot find answers from the code in this repository:

As I read from the paper, the pretraining data is 1.6 billion tweets, each one labeled with an emoji. Then when splitting the dataset, you sampled 10k tweets for each emoji type for both the validation set and test set, therefore each set has 640k different samples. Later you upsampled the remaining data, making sure each emoji type has the same number of samples. Does it mean you simply repeated the tweets from infrequent emojis? In that case, your training set somehow became tens of billions of samples, instead of 1.6 billions. Did I understand it correctly? 

Because from what I have learned, repeated samples could lead to poor improvement with minibatches, therefore I'm curious whether I understood your method correctly or not. Thanks in advance!

**Comment by bfelbo** [comment]
Yes, we repeated the infrequent emoji tweets. Your concern is justified, but it was not an issue in our case due to the large amounts of tweets even for our most infrequent emoji (see the table in the paper).

On a technical note, I would suggest that you use an upsampling generator that does this upsampling on-the-fly rather than doing upsampling once and then storing the dataset. This works better and also prevents you from having to store a giant dataset with many billions of tweets.
